### PR TITLE
avoid deadlock when sql engine if full with coordinators

### PR DESCRIPTION
**Pull request by dorinhogea**
Ported https://github.com/bloomberg/comdb2/pull/2841 to master.

Signed-off-by: Dorin Hogea <dhogea@bloomberg.net>

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/769ff4ef/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/769ff4ef/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/769ff4ef/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2842'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/108ee107/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>1/406 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_timepart_logicalsc_generated</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/108ee107/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/108ee107/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2842'>Resubmit</a></p>

File: db/db_tunables.h [hunk 1]
```diff
@@ -1664,13 +1664,19 @@ REGISTER_TUNABLE("dohast_verbose",
 REGISTER_TUNABLE("dohsql_max_queued_kb_highwm",
                  "Maximum shard queue size, in KB; shard sqlite will pause "
                  "once queued bytes limit is reached.",
-                 TUNABLE_INTEGER, &gbl_dohsql_max_queued_kb_highwm, 0, NULL,
+                 TUNABLE_INTEGER, &gbl_dohsql_max_queued_kb_highwm, 10000, NULL,
```
Comment (mponomar) [inline db/db_tunables.h:1667]: This is flags, not default (though this would be the perfect place to specify defaults...)

File: db/db_tunables.h [hunk 2]
```diff
@@ -1664,13 +1664,19 @@ REGISTER_TUNABLE("dohast_verbose",
...
     "dohsql_max_threads",
     "Maximum number of parallel threads, otherwise run sequential.",
-    TUNABLE_INTEGER, &gbl_dohsql_max_threads, 0, NULL, NULL, NULL, NULL);
+    TUNABLE_INTEGER, &gbl_dohsql_max_threads, 8, NULL, NULL, NULL, NULL);
+
+REGISTER_TUNABLE(
+    "dohsql_pool_thread_slack",
+    "Forbid parallel sql coordinators from running on this many sql engines"
+    " (if 0, defaults to 1).",
+    TUNABLE_INTEGER, &gbl_dohsql_pool_thr_slack, 1, NULL, NULL, NULL, NULL);
```
Comment (mponomar) [inline db/db_tunables.h:1679]: Also flags

File: db/dohsql.c [hunk 3]
```diff
@@ -1281,6 +1330,8 @@ int dohsql_end_distribute(struct sqlclntstate *clnt, struct reqlogger *logger)
     clnt->conns = NULL;
     free(conns);
 
+    _rem_parallel_load();
```
Comment (mponomar) [inline db/dohsql.c:1333]: There's a couple of early returns that don't _rem_parallel_load

Comment (dorinhogea) [reply db/dohsql.c:1333]: Only one return, if failed to run recover_deadlock.  That is probably unrecoverable though. We'll add the call to _rem_parallel_load anyway

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/94afc788/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/94afc788/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/94afc788/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2842'>Resubmit</a></p>

**Review by adizaimi — APPROVED** [review]
LGTM.

**Comment by dorinhogea** [comment]
TYA!

**Pull request merged**